packages/common/src/codegen/utils/parseSystem.ts

```diff
@@ -10,32 +12,34 @@ export function parseSystem(
   source: string,
   contractName: string,
 ): undefined | { contractType: "contract" | "abstract" } {
-  const ast = parse(source);
-  const contractNode = findContractNode(ast, contractName);
-  if (!contractNode) return;
+  const parser = Parser.create(LanguageFacts.latestVersion());
+  const root = parser.parseFileContents(source).createTreeCursor();
 
-  const contractType = contractNode.kind;
-  // skip libraries and interfaces
-  // we allow abstract systems here so that we can create system libraries from them but without deploying them
-  if (contractType !== "contract" && contractType !== "abstract") return;
+  const contractCursor = findContractOrInterfaceNode(root, contractName);
+  if (!contractCursor) return;
 
-  const isSystem = ((): boolean => {
-    // if using the System suffix, assume its a system
-    if (contractName.endsWith("System") && contractName !== baseSystemName) return true;
+  assertNonterminalNode(contractCursor.node);
```

Comment:
I'd expect this to be a postcondition of `findCotnractOrInterfaceNode`, and have the check there instead.

Reply:
This is needed to do type narrowing on the `node` to be able to use it later in the `new` statement. The post-condition is not important. So, moving the the assertion into `findContractOrInterfaceNode` would not work.

Reply:
Ah, but wouldn't a `let contractNode = contractCursor.node.asNonterminalNode()` be the most appropriate?

Reply:
Hmm... `contractCursor.node.asNonterminalNode()` types to `undefined | NonterminalNode` so we'd need to type-narrow it again. It's my understanding that `assertTerminalNode` and `assertNonterminalNode` are built specifically for this type of use cases.

Reply:
You're totally right. Perhaps we need a method for it?